# Listed Data EDA (Week1) 
Author : Yoora Choi

In [68]:
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

In [69]:
# Import listed data (2024.01-2026.03)
listed = pd.read_csv("CRMLSListing_merged.csv")

/var/folders/d2/04b97b6n77b6vr8qn0rd3q340000gn/T/ipykernel_51587/959516333.py:2: DtypeWarning: Columns (2,43) have mixed types. Specify dtype option on import or set low_memory=False.
  listed = pd.read_csv("CRMLSListing_merged.csv")


## 1.Data overview

In [70]:
# Check data structure
listed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 853812 entries, 0 to 853811
Data columns (total 84 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   OriginalListPrice             850502 non-null  float64
 1   ListingKey                    853812 non-null  int64  
 2   ListAgentEmail                780379 non-null  object 
 3   CloseDate                     255634 non-null  object 
 4   ClosePrice                    232761 non-null  float64
 5   ListAgentFirstName            848796 non-null  object 
 6   ListAgentLastName             853736 non-null  object 
 7   Latitude                      741864 non-null  float64
 8   Longitude                     742568 non-null  float64
 9   UnparsedAddress               851654 non-null  object 
 10  PropertyType                  853812 non-null  object 
 11  LivingArea                    747336 non-null  float64
 12  ListPrice                     851677 non-nul

In [86]:
# check the data structure
listed.head(5)

,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,FireplacesTotal,AssociationFeeFrequency,AboveGradeFinishedArea,ListingKeyNumeric,MLSAreaMajor,TaxAnnualAmount,CountyOrParish,PropertyType.1,MlsStatus,ElementarySchool,ListAgentFirstName.1,AttachedGarageYN,ParkingTotal,BuilderName,PropertySubType,LotSizeAcres,SubdivisionName,BuyerOfficeAOR,YearBuilt,DaysOnMarket.1,BuyerAgencyCompensationType,StreetNumberNumeric,LivingArea.1,ListingId,BathroomsTotalInteger,City,BuyerAgencyCompensation,TaxYear,BuildingAreaTotal,BedroomsTotal,ContractStatusChangeDate,Longitude.1,ElementarySchoolDistrict,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,BelowGradeFinishedArea,BusinessType,Latitude.1,ListPrice.1,StateOrProvince,CoveredSpaces,MiddleOrJuniorSchool,FireplaceYN,Stories,HighSchool,Levels,ListAgentLastName.1,CloseDate.1,LotSizeDimensions,LotSizeArea,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,BuyerOfficeName.1,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,UnparsedAddress.1
0,90000.0,1075010398,miriamlara03@gmail.com,NaN,NaN,Miriam,Lara,34.097939,-117.909653,1045 N Azusa 61,ManufacturedInPark,960.0,90000.0,0,CENTURY 21 CITRUS REALTY INC,NaN,NaN,Miriam Lara,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1075010398,614 - Covina,NaN,Los Angeles,ManufacturedInPark,Active,NaN,Miriam,False,4.0,NaN,NaN,NaN,NaN,NaN,1960.0,0,Item,1045.0,960.0,CV24093347,1.0,Covina,2500.0,NaN,NaN,4.0,2024-01-29,-117.909653,NaN,NaN,NaN,2024-01-29,NaN,NaN,34.097939,90000.0,CA,NaN,NaN,NaN,1.0,NaN,One,Lara,NaN,NaN,NaN,NaN,NaN,2.0,Covina Valley Unified,91722,NaN,0.0,NaN,NaN,1045 N Azusa 61
1,1500000.0,1074974457,janelle@judsonre.com,NaN,NaN,Janelle,Judson,33.121241,-117.081614,NaN,CommercialSale,NaN,1500000.0,0,Judson Real Estate,NaN,Judson Real Estate,Janelle Judson,Lynn,Judson,NaN,NaN,NaN,NaN,NaN,NaN,1074974457,92025 - Escondido,NaN,San Diego,CommercialSale,Active,NaN,Janelle,NaN,0.0,NaN,Retail,NaN,NaN,NaN,1940.0,0,Item1,NaN,NaN,NDP2403932,0.0,Escondido,2.0,NaN,3000.0,NaN,2024-01-11,-117.081614,NaN,NaN,NaN,2024-01-11,NaN,NaN,33.121241,1500000.0,CA,NaN,NaN,NaN,NaN,NaN,NaN,Judson,NaN,NaN,3052.0,NaN,NaN,NaN,NaN,92025,NaN,0.0,NaN,NaN,NaN
2,1340000.0,1074973329,haleh360@Gmail.com,NaN,NaN,Haleh,Dowlatshahi,34.052207,-118.408445,2220 Avenue Of The Stars 2704,Residential,1301.0,1340000.0,127,Rodeo Realty- Brentwood,NaN,NaN,Haleh Dowlatshahi,NaN,NaN,NaN,NaN,NaN,NaN,Monthly,NaN,1074973329,C05 - Westwood - Century City,NaN,Los Angeles,Residential,Pending,NaN,Haleh,NaN,0.0,NaN,Condominium,4.0831,NaN,NaN,1964.0,127,Item1,2220.0,1301.0,24389291,2.0,Los Angeles,2.0,NaN,1301.0,2.0,2024-05-07,-118.408445,NaN,NaN,2024-05-07,2024-01-01,NaN,NaN,34.052207,1340000.0,CA,NaN,NaN,False,1.0,NaN,One,Dowlatshahi,NaN,NaN,177861.0,NaN,False,NaN,NaN,90067,NaN,2105.0,177861.0,NaN,2220 Avenue Of The Stars 2704
3,2500000.0,1074954552,Reneechen@yourhomesoldguaranteed.com,NaN,NaN,Renee,Chen,33.496363,-117.691677,16 Palisades,Residential,2788.0,2500000.0,1,Your Home Sold Guaranteed Realty,NaN,NaN,Renee Chen,NaN,NaN,NaN,NaN,NaN,NaN,Monthly,NaN,1074954552,LNSLT - Salt Creek,NaN,Orange,Residential,Active,NaN,Renee,True,3.0,NaN,SingleFamilyResidence,0.1217,Warmington Homes (BB) (BBW),NaN,1994.0,1,Item1,16.0,2788.0,AR24091628,3.0,Laguna Niguel,2.0,NaN,NaN,4.0,2024-01-24,-117.691677,NaN,NaN,NaN,2024-01-24,NaN,NaN,33.496363,2500000.0,CA,NaN,NaN,True,2.0,NaN,Two,Chen,NaN,NaN,5300.0,0.0,False,3.0,Capistrano Unified,92677,NaN,254.0,5300.0,NaN,16 Palisades
4,3150000.0,1074936537,anader@dppre.com,NaN,NaN,Margaret,Nader,34.119345,-118.111254,1615 Waverly Road,Residential,3250.0,3150000.0,1,COMPASS,NaN,NaN,Margaret Nader,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1074936537,655 - San Marino,

In [72]:
# Summary statistics for numerical variable
listed.describe()

,OriginalListPrice,ListingKey,ClosePrice,Latitude,Longitude,LivingArea,ListPrice,DaysOnMarket,FireplacesTotal,AboveGradeFinishedArea,ListingKeyNumeric,TaxAnnualAmount,ParkingTotal,LotSizeAcres,YearBuilt,DaysOnMarket.1,StreetNumberNumeric,LivingArea.1,BathroomsTotalInteger,BuyerAgencyCompensation,TaxYear,BuildingAreaTotal,BedroomsTotal,Longitude.1,ElementarySchoolDistrict,BelowGradeFinishedArea,Latitude.1,ListPrice.1,CoveredSpaces,Stories,LotSizeArea,MainLevelBedrooms,GarageSpaces,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict
count,8.505020e+05,8.538120e+05,2.327610e+05,741864.000000,742568.000000,7.473360e+05,8.516770e+05,853812.000000,0.0,0.0,8.538120e+05,6.093000e+03,8.014550e+05,7.722370e+05,784342.000000,853812.000000,8.500390e+05,7.473360e+05,783863.000000,123519.000000,868.000000,1.369110e+05,750731.000000,742568.000000,0.0,3421.000000,741864.000000,8.516770e+05,0.0,632306.000000,7.760150e+05,394763.000000,691686.000000,577832.000000,7.749420e+05,0.0
mean,1.051202e+06,1.105021e+09,7.811907e+05,34.519494,-118.364673,3.159195e+03,9.842388e+05,19.128794,NaN,NaN,1.105021e+09,3.061311e+04,6.431168e+00,5.507730e+02,1978.399809,19.128794,2.826086e+04,3.159195e+03,2.498272,175.053222,2023.326037,3.343991e+03,3.132479,-118.364673,NaN,96.899889,34.519494,9.842388e+05,NaN,1.367672,6.648095e+04,1.933398,1.756370,205.317739,5.623964e+05,NaN
std,6.389677e+06,2.997224e+07,3.458990e+06,1.885883,3.887197,1.051786e+06,2.167899e+06,28.843245,NaN,NaN,2.997224e+07,9.655613e+05,2.906784e+03,4.214830e+05,28.041748,28.843245,2.042188e+06,1.051786e+06,3.034337,2053.693620,2.653963,1.135116e+04,1.689056,3.887197,NaN,497.457736,1.885883,2.167899e+06,NaN,0.482172,2.374090e+06,5.600212,3.940488,2060.873314,2.310913e+07,NaN
min,0.000000e+00,4.216549e+08,0.000000e+00,-117.472493,-159.523315,0.000000e+00,0.000000e+00,-58.000000,NaN,NaN,4.216549e+08,0.000000e+00,-1.430000e+02,0.000000e+00,1776.000000,-58.000000,0.000000e+00,0.000000e+00,0.000000,0.000000,2004.000000,0.000000e+00,0.000000,-159.523315,NaN,0.000000,-117.472493,0.000000e+00,NaN,1.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000e+00,NaN
25%,8.500000e+04,1.077490e+09,5.000000e+03,33.757792,-118.517241,1.182000e+03,8.500000e+04,4.000000,NaN,NaN,1.077490e+09,1.190000e+02,1.000000e+00,1.205000e-01,1960.000000,4.000000,7.650000e+02,1.182000e+03,2.000000,2.000000,2023.000000,1.219000e+03,2.000000,-118.517241,NaN,0.000000,33.757792,8.500000e+04,NaN,1.000000,4.936000e+03,1.000000,1.000000,0.000000,5.227000e+03,NaN
50%,6.399000e+05,1.104373e+09,5.650000e+05,34.051492,-118.091547,1.601000e+03,6.350000e+05,10.000000,NaN,NaN,1.104373e+09,6.282000e+03,2.000000e+00,1.722000e-01,1979.000000,10.000000,2.758000e+03,1.601000e+03,2.000000,2.500000,2024.000000,1.870000e+03,3.000000,-118.091547,NaN,0.000000,34.051492,6.350000e+05,NaN,1.000000,7.227000e+03,2.000000,2.000000,0.000000,7.500000e+03,NaN
75%,1.150000e+06,1.127482e+09,1.025000e+06,34.302730,-117.289035,2.236000e+03,1.150000e+06,22.000000,NaN,NaN,1.127482e+09,1.962000e+04,3.000000e+00,3.846000e-01,2001.000000,22.000000,1.080800e+04,2.236000e+03,3.000000,2.500000,2024.000000,3.209000e+03,4.000000,-117.289035,NaN,0.000000,34.302730,1.150000e+06,NaN,2.000000,1.368850e+04,3.000000,2.000000,311.000000,1.655300e+04,NaN
max,1.390000e+09,1.156489e+09,8.200000e+08,737.000000,392.000000,9.090909e+08,4.436503e+08,2539.000000,NaN,NaN,1.156489e+09,7.446000e+07,2.593669e+06,3.702600e+08,2028.000000,2539.000000,7.212710e+08,9.090909e+08,2208.000000,200000.000000,2026.000000,1.200000e+06,149.000000,392.000000,NaN,17424.000000,737.000000,4.436503e+08,NaN,2.000000,9.187423e+08,3333.000000,600.000000,968348.000000,7.638115e+09,NaN


In [73]:
# Summary statistics for non-numerical variable
listed.describe(include = 'object')

,ListAgentEmail,CloseDate,ListAgentFirstName,ListAgentLastName,UnparsedAddress,PropertyType,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,AssociationFeeFrequency,MLSAreaMajor,CountyOrParish,PropertyType.1,MlsStatus,ElementarySchool,ListAgentFirstName.1,AttachedGarageYN,BuilderName,PropertySubType,SubdivisionName,BuyerOfficeAOR,BuyerAgencyCompensationType,ListingId,City,ContractStatusChangeDate,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,BusinessType,StateOrProvince,MiddleOrJuniorSchool,FireplaceYN,HighSchool,Levels,ListAgentLastName.1,CloseDate.1,LotSizeDimensions,NewConstructionYN,HighSchoolDistrict,PostalCode,BuyerOfficeName.1,UnparsedAddress.1
count,780379,255634,848796,853736,851654,853812,853812,245209,170949,853584,170318,170805,246797,246636,247545,246411,757389,853811,853812,853812,79296,848796,562715,29299,749640,285106,233426,123538,853812,852850,846890,19685,368069,853812,5977,853746,79387,731256,104344,689733,853736,255634,48707,739086,514356,853604,245209,851654
unique,104556,918,19205,41808,713050,8,24204,16940,7083,98915,8412,17341,89309,15875,32749,4,1153,63,8,5,1446,19205,2,4601,34,21838,66,3,853662,1324,848,3191,831,821,1911,41,679,2,534,19,41808,918,17394,2,454,5298,16940,713050
top,info@nikolaleasing.com,2024-03-29,Michael,Lee,0 0,Residential,Compass,Compass,Compass,Niko Deleon,Michael,Myatt,NONMEMBER,NONE,NONE,Monthly,699 - Not Defined,Los Angeles,Residential,Active,Other,Michael,True,Lennar,SingleFamilyResidence,Other,OrangeCounty,Item1,PTP2403263,Los Angeles,2024-04-30,Michael,2024-05-01,2025-05-01,Restaurant,CA,Other,True,Other,One,Lee,2024-03-29,50x135,False,Other,92253,Compass,0 0
freq,2146,1375,10895,5152,147,540612,53400,17135,19359,2173,2696,1134,3674,4937,4873,230622,84997,266159,540612,471243,14994,10895,432615,2569,478770,7906,22421,113085,5,68700,3242,259,1227,2648,839,853449,11274,439523,4690,395035,5152,1375,896,708578,55178,5310,17135,147


## 2. Missing data check(null)

In [74]:
# null count
null_count = listed.isnull().sum()

# null percentage (%)
null_percent = (listed.isnull().sum() / len(listed)) * 100

# data frame
null_summary = pd.DataFrame({
    'null_count': null_count,
    'null_percent': null_percent
}).sort_values(by='null_percent', ascending=False)

null_summary['null_percent'] = null_summary['null_percent'].round(2)
null_summary

,null_count,null_percent
CoveredSpaces,853812,100.00
MiddleOrJuniorSchoolDistrict,853812,100.00
ElementarySchoolDistrict,853812,100.00
AboveGradeFinishedArea,853812,100.00
FireplacesTotal,853812,100.00
TaxYear,852944,99.90
BelowGradeFinishedArea,850391,99.60
BusinessType,847835,99.30
TaxAnnualAmount,847719,99.29
CoBuyerAgentFirstName,834127,97.69


## 3. Check duplicates

### 3.1 duplicated columns

In [75]:
# check duplicate columns
pairs = []

for col in listed.columns:
    if '.' in col:
        base_col = col.split('.')[0]
        if base_col in listed.columns:
            pairs.append((base_col, col))

pairs

[('PropertyType', 'PropertyType.1'),
 ('ListAgentFirstName', 'ListAgentFirstName.1'),
 ('DaysOnMarket', 'DaysOnMarket.1'),
 ('LivingArea', 'LivingArea.1'),
 ('Longitude', 'Longitude.1'),
 ('Latitude', 'Latitude.1'),
 ('ListPrice', 'ListPrice.1'),
 ('ListAgentLastName', 'ListAgentLastName.1'),
 ('CloseDate', 'CloseDate.1'),
 ('BuyerOfficeName', 'BuyerOfficeName.1'),
 ('UnparsedAddress', 'UnparsedAddress.1')]

In [76]:
# Check if duplicate column pairs have identical values row-by-row
for base, dup in pairs:
    same_ratio = (listed[base] == listed[dup]).mean()
    print(f"{base} vs {dup}: {same_ratio:.2f}")

PropertyType vs PropertyType.1: 1.00
ListAgentFirstName vs ListAgentFirstName.1: 0.99
DaysOnMarket vs DaysOnMarket.1: 1.00
LivingArea vs LivingArea.1: 0.88
Longitude vs Longitude.1: 0.87
Latitude vs Latitude.1: 0.87
ListPrice vs ListPrice.1: 1.00
ListAgentLastName vs ListAgentLastName.1: 1.00
CloseDate vs CloseDate.1: 0.30
BuyerOfficeName vs BuyerOfficeName.1: 0.29
UnparsedAddress vs UnparsedAddress.1: 1.00


### 3.2 duplicated rows

In [77]:
# check completely duplicated rows
listed[listed.duplicated(keep=False)]

,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,FireplacesTotal,AssociationFeeFrequency,AboveGradeFinishedArea,ListingKeyNumeric,MLSAreaMajor,TaxAnnualAmount,CountyOrParish,PropertyType.1,MlsStatus,ElementarySchool,ListAgentFirstName.1,AttachedGarageYN,ParkingTotal,BuilderName,PropertySubType,LotSizeAcres,SubdivisionName,BuyerOfficeAOR,YearBuilt,DaysOnMarket.1,BuyerAgencyCompensationType,StreetNumberNumeric,LivingArea.1,ListingId,BathroomsTotalInteger,City,BuyerAgencyCompensation,TaxYear,BuildingAreaTotal,BedroomsTotal,ContractStatusChangeDate,Longitude.1,ElementarySchoolDistrict,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,BelowGradeFinishedArea,BusinessType,Latitude.1,ListPrice.1,StateOrProvince,CoveredSpaces,MiddleOrJuniorSchool,FireplaceYN,Stories,HighSchool,Levels,ListAgentLastName.1,CloseDate.1,LotSizeDimensions,LotSizeArea,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,BuyerOfficeName.1,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,UnparsedAddress.1


In [78]:
listed.duplicated().sum()

np.int64(0)

In [79]:
# 컬럼명 확인
print(listed.columns[2])
print(listed.columns[43])

ListAgentEmail
BuyerAgencyCompensationType


In [80]:
print(listed['ListAgentEmail'].unique())
print(listed['BuyerAgencyCompensationType'].unique())

['miriamlara03@gmail.com' 'janelle@judsonre.com' 'haleh360@Gmail.com' ...
 'FrankHundley@gmail.com' 'bunnysellssandiego@gmail.com'
 'karen.silvas@compass.com']
['Item' 'Item1' 'SeeRemarks' nan]


## Data Analysis Summary (CRMLSListing data (2024.01-2026.03))

### 1. Basic Information
- **Total Rows:** 853,812
- **Total Columns:** 84
- **Data Period:** January 2024 – March 2026

---

### 2. Column Structure
- **Data Types:** float64(32), int64(4), object(48)

---

### 3. Missing Data
#### Completely Empty Columns (100% Null) → Consider Dropping
- `CoveredSpaces`, `MiddleOrJuniorSchoolDistrict`, `ElementarySchoolDistrict`, `AboveGradeFinishedArea`, `FireplacesTotal`

#### High Missing Rate (80–100%)
- `TaxYear` (99.90%), `BelowGradeFinishedArea` (99.60%), `BusinessType` (99.30%), `TaxAnnualAmount` (99.29%), `CoBuyerAgentFirstName` (97.69%), `BuilderName` (96.57%), `LotSizeDimensions` (94.30%), `ElementarySchool` (90.71%), `MiddleOrJuniorSchool` (90.70%), `HighSchool` (87.78%), `BuyerAgencyCompensation` (85.53%), `BuyerAgencyCompensationType` (85.53%), `BuildingAreaTotal` (83.96%)

#### High Missing Rate (70–80%)
- `CoListAgentFirstName` (80.05%), `CoListAgentLastName` (80.00%), `CoListOfficeName` (79.98%), `ClosePrice` (72.74%), `BuyerOfficeAOR` (72.66%), `BuyerOfficeName` (71.28%), `BuyerOfficeName.1` (71.28%), `AssociationFeeFrequency` (71.14%), `BuyerAgentFirstName` (71.11%), `BuyerAgentMlsId` (71.09%), `BuyerAgentLastName` (71.01%), `CloseDate` (70.06%), `CloseDate.1` (70.06%)
- → Buyer-related columns are missing for non-closed listings, which is expected

#### Moderate Missing Rate (10–70%)
- `SubdivisionName` (66.61%), `PurchaseContractDate` (56.89%), `MainLevelBedrooms` (53.76%), `HighSchoolDistrict` (39.76%), `AttachedGarageYN` (34.09%), `AssociationFee` (32.32%), `Stories` (25.94%), `Levels` (19.22%), `GarageSpaces` (18.99%), `FireplaceYN` (14.35%), `NewConstructionYN` (13.44%), `Latitude` (13.11%), `Latitude.1` (13.11%), `Longitude` (13.03%), `Longitude.1` (13.03%), `LivingArea` (12.47%), `LivingArea.1` (12.47%), `PropertySubType` (12.20%), `BedroomsTotal` (12.07%), `MLSAreaMajor` (11.29%)

#### Low Missing Rate (0–10%)
- `LotSizeAcres` (9.55%), `LotSizeSquareFeet` (9.24%), `LotSizeArea` (9.11%), `ListAgentEmail` (8.60%), `BathroomsTotalInteger` (8.19%), `YearBuilt` (8.14%), `ParkingTotal` (6.13%), `ContractStatusChangeDate` (0.81%), `ListAgentFirstName` (0.59%), `ListAgentFirstName.1` (0.59%), `StreetNumberNumeric` (0.44%), `OriginalListPrice` (0.39%), `UnparsedAddress` (0.25%), `UnparsedAddress.1` (0.25%), `ListPrice` (0.25%), `ListPrice.1` (0.25%), `City` (0.11%), `ListAgentFullName` (0.03%), `PostalCode` (0.02%), `ListAgentLastName` (0.01%), `ListAgentLastName.1` (0.01%), `StateOrProvince` (0.01%), `CountyOrParish` (0.00%)

#### No Missing Values
- `DaysOnMarket`, `DaysOnMarket.1`, `PropertyType`, `PropertyType.1`, `ListingKey`, `ListingKeyNumeric`, `ListOfficeName`, `MlsStatus`, `ListingId`, `ListingContractDate`

---

### 4. Duplicate Columns (11 Pairs Found)

| Same Value Ratio | Column Pairs |
|-----------------|--------------|
| 100% | `PropertyType` / `PropertyType.1`, `DaysOnMarket` / `DaysOnMarket.1`, `ListPrice` / `ListPrice.1`, `ListAgentLastName` / `ListAgentLastName.1`, `UnparsedAddress` / `UnparsedAddress.1` |
| ~99% | `ListAgentFirstName` / `ListAgentFirstName.1` |
| ~87–88% | `Latitude` / `Latitude.1`, `Longitude` / `Longitude.1`, `LivingArea` / `LivingArea.1` |
| **~29–30%** | **`CloseDate` / `CloseDate.1`, `BuyerOfficeName` / `BuyerOfficeName.1`** ← Values differ, requires investigation |

---

### 5. Duplicate Rows
- **Completely duplicated rows: 0** 



### 6. Key MLS Fields – Statistical Summary

In [81]:
# Numerical key fields summary
key_num_cols = ['ListPrice', 'ClosePrice', 'LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger', 'DaysOnMarket']
listed[key_num_cols].describe().round(2)

,ListPrice,ClosePrice,LivingArea,BedroomsTotal,BathroomsTotalInteger,DaysOnMarket
count,8.516770e+05,2.327610e+05,7.473360e+05,750731.00,783863.00,853812.00
mean,9.842388e+05,7.811907e+05,3.159190e+03,3.13,2.50,19.13
std,2.167899e+06,3.458990e+06,1.051786e+06,1.69,3.03,28.84
min,0.000000e+00,0.000000e+00,0.000000e+00,0.00,0.00,-58.00
25%,8.500000e+04,5.000000e+03,1.182000e+03,2.00,2.00,4.00
50%,6.350000e+05,5.650000e+05,1.601000e+03,3.00,2.00,10.00
75%,1.150000e+06,1.025000e+06,2.236000e+03,4.00,3.00,22.00
max,4.436503e+08,8.200000e+08,9.090909e+08,149.00,2208.00,2539.00


**Data Quality Issues Detected**
- **ListPrice / ClosePrice / LivingArea:** Min = 0 and extremely high max values suggest outliers or data entry errors which, requires cleaning
- **DaysOnMarket:** Negative min value (-58) is invalid, which requires investigation
- **BedroomsTotal:** Max = 149 is likely an error
- **BathroomsTotalInteger:** Max = 2,208 is clearly an error

In [82]:
# Categorical key fields summary
key_cat_cols = ['PropertyType', 'ListOfficeName']
listed[key_cat_cols].describe()

,PropertyType,ListOfficeName
count,853812,853812
unique,8,24204
top,Residential,Compass
freq,540612,53400


In [83]:
# Categorical key fields - value counts
for col in key_cat_cols:
    print(f"\n--- {col} ---")
    print(listed[col].value_counts())


--- PropertyType ---
PropertyType
Residential            540612
ResidentialLease       178432
Land                    56352
ResidentialIncome       31683
ManufacturedInPark      24540
CommercialSale          11721
CommercialLease          7740
BusinessOpportunity      2732
Name: count, dtype: int64

--- ListOfficeName ---
ListOfficeName
Compass                                                               53400
Coldwell Banker Realty                                                35402
Keller Williams Realty                                                12642
Berkshire Hathaway HomeServices California Properties                 10641
First Team Real Estate                                                 8852
eXp Realty of California Inc                                           8259
Real Broker                                                            7979
Equity Union                                                           7781
Intero Real Estate Services                         

### 7. Next Steps
1. Drop 100% null columns: `CoveredSpaces`, `MiddleOrJuniorSchoolDistrict`, `ElementarySchoolDistrict`, `AboveGradeFinishedArea`, `FireplacesTotal`
2. Drop `.1` duplicate columns, but first investigate why `CloseDate` and `BuyerOfficeName` show only ~30% match
3. Determine how to handle missing values in Buyer-related columns
4. Reload CSV with explicit dtype specified for mixed-type columns: `ListAgentEmail`, `BuyerAgencyCompensationType`
5. Fix data types for the following columns:
   - `CloseDate`, `CloseDate.1`, `PurchaseContractDate`, `ListingContractDate`, `ContractStatusChangeDate` : object → datetime
   - `YearBuilt`, `BedroomsTotal`, `BathroomsTotalInteger`, `Stories`, `GarageSpaces`, `ParkingTotal`, `MainLevelBedrooms` : float64 → int64 (after handling nulls)
   - `AttachedGarageYN`, `FireplaceYN`, `NewConstructionYN` : object → bool